git/fetchpack.go

```diff
@@ -446,11 +446,13 @@ func (p *packProtocolReader) Read(buf []byte) (int, error) {
 				}
 				goto sidebandRead
 			case sidebandErrChannel:
-				n, err := io.ReadFull(p.conn, buf[:size-5])
+				msgbuf := make([]byte, size)
```

Comment:
doesn't getting rid of the -5 mean you're overreading? The subtraction is intended to compensate for the length prefix in the pack protocol that was already read.

It's difficult to say what the appropriate value for "n" to return here is.. I think changing the return to 0 probably makes sense and makes the error sideband channel behave more like the status sideband channel (which goes back and does another read, so we never do anything with the n from its read), because implicitly a Read from a pack protocol reader is trying to read the packfile, but the error sideband channel is for remote errors (which is why we return here, unlike sidebandChannel case). Why is it generating a panic? Is the caller not checking the err return? And are the tests that are failing explicitly trying to test for bad fetches?

Reply:
One panic was right here because the buf was smaller than size-5. The next happened when I allocated a buffer that was only size-5 and the panic was in io.ReadFull itself. Using size seems to work. I'm not sure why based on what you're saying.

It's hard to pinpoint what exactly is going on with t5510-fetch because there are a lot of things failing at the moment. I think I identified that the git-upload-pack process aborts because it has decided that the repo is corrupt. I figured that we shouldn't be panicking on our end at least.

Reply:
Can you add an if statement to compare size-5 to len(buf) before passing the slice to ReadFull?

Reply:
I added back the size-5 for the buffer that is allocated. It seems that I was confused about the panic inside io.ReadFull.